### ISSUE TITLE
Better way to handle the timeout when using FileWillRenameEvent.waitUntil

### ISSUE THREAD

**Issue by testforstephen** [microsoft/vscode#111208]
When refactoring Java files via File Explorer, such as moving the files to another package or renaming it to a new package name, we need to calculate the changes for the refactoring before the actual renaming occurs. For such a refactoring, we listen to the onWillRenameFiles event and use the FileWillRenameEvent.waitUntil() method to return the edit as result.

One limitation of waitUntil is that it defaults to a 5-second timeout. When it times out, the client is forced to rename the files to a new location regardless of whether the refactoring is complete or not. When refactoring a large project, this may cause the refactoring to fail silently. Although VS Code provides a user setting to change the timeout value, i believe most of users don't have the habit to find and customize the setting, they probably take this as the product stability issue.

One idea in my mind is to ask user whether to continue waiting the refactoring job instead of timeout silently.

**Comment by jrieken** [comment]
> One idea in my mind is to ask user whether to continue waiting the refactoring job instead of timeout silently.

I like that. We could show a progress notification with an option to cancel the operation

**Comment by jrieken** [comment]
@testforstephen fyi I have pushed a change that does two things

1. the default value for `files.participants.timeout` is now 60 seconds
2. after a short while a progress notification is shown which allows to cancel onWill-event handler

**Comment by jrieken** [comment]
To _verify_

* have an extension snippet as show below
* rename/move a file
* make sure that you see a progress message after a shorting while
* make sure that pressing cancel actually cancels the operation, e.g the text edit isn't applied


```ts

 vscode.workspace.onWillRenameFiles(e => {

        const [first] = e.files;

        e.waitUntil(vscode.workspace.openTextDocument(first.oldUri).then(async doc => {
            await new Promise(resolve => setTimeout(resolve, 10000))
            const edit = new vscode.WorkspaceEdit();
            edit.insert(first.oldUri, new vscode.Position(0, 0), `I HAVE BEEN MOVED ${new Date().toDateString()}\n`);
            return edit;
        }));
    })
```

**Comment by testforstephen** [comment]
Thanks for the fix.

To verify this in the Java extension, I changed the implementation of the rename refactoring by using the `workspace.onWillRenameFiles` handler, and now after I rename fileA.java to fileB.java, one undo will restore the refactor and the filename.

But I also found two potential problems about the willRenameFiles handler.

- The UI state is in chaos if the returned edit enables the preview flag.

If `FileWillRenameEvent.waitUntil()` returns a workspace edit with the preview flag enabled (e.g. add `needsConfirmation` properties), the UI state will be in chaos. See the screenshot below. It shows REFACTOR PREVIEW panel, but the file explorer is frozen and also shows the running notification.
![image](https://user-images.githubusercontent.com/14052197/101131662-a2ad3800-3640-11eb-91af-86e223084ba3.png)

- DnD multiple files to a new location, it returns multiple separate events instead of a single FileWillRenameEvent.

If I drag and drop multiple files into a new location at once, the file event API `workspace.onWillRenameFiles` will report multiple separate FileWillRenameEvent for each file to be moved. If i choose to use FileWillRenameEvent.waitUntil to return the refactored edit, that means each event will return a separate edit and the client will call the applyEdit operation multiple times. Undo cannot roll back the text changes correctly. It's better to report a single file event instead of separate events when DnD.  Opened a new issue for this https://github.com/microsoft/vscode/issues/111867.
